File: 21. Merge Two Sorted Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        // easy situation
        if (list1 == null) {
            return list2;
        }
        if (list2 == null) {
            return list1;
        }
        // beginning
        ListNode beginning;
        // going through the nodes
        ListNode cursor;
        // figure out the beginning
        if (list1.val < list2.val) {
            cursor = beginning = new ListNode(list1.val);
            list1 = list1.next;
        } else {
            cursor = beginning = new ListNode(list2.val);
            list2 = list2.next;
        }
        // Loop until you hit the end of the list
        while (list1 != null && list2 != null) {
            if (list1.val < list2.val) {
                cursor.next = new ListNode(list1.val);
                list1 = list1.next;
            } else {
                cursor.next = new ListNode(list2.val);
                list2 = list2.next;
            }
            cursor = cursor.next;
        }
        // dump the remaining in list 1
        while (list1 != null) {
            cursor.next = new ListNode(list1.val);
            list1 = list1.next;
            cursor = cursor.next;
        }
        // dump the remaining in list 2
        while (list2 != null) {
            cursor.next = new ListNode(list2.val);
            list2 = list2.next;
            cursor = cursor.next;
        }
        return beginning;
    }
    
}

File: 21. Merge Two Sorted Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:

        list3 = ListNode(100)
        original = list3
        while(list1 is not None or list2 is not None):
            list3.next = ListNode(0)
            list3 = list3.next
            if(list1 is None): 
                list3.val = list2.val
                list2 = list2.next
            elif(list2 is None):
                list3.val = list1.val
                list1 = list1.next
            else:
                if(list1.val <= list2.val):
                    list3.val = list1.val
                    list1 = list1.next
                else:
                    list3.val = list2.val
                    list2 = list2.next
        return original.next

        #leng = len(list1) + len(list2)
        #for x in range(leng):
